.lintrunner.toml

```diff
@@ -168,3 +170,44 @@ command = [
     '@{{PATHSFILE}}'
 ]
 is_formatter = true
+
+[[linter]]
+code = 'CLANGFORMAT'
+include_patterns = [
+    '**/*.h',
+    '**/*.c',
+    '**/*.hh',
+    '**/*.cc',
+    '**/*.hpp',
+    '**/*.cpp',
+    '**/*.m',
+    '**/*.mm',
+    '**/*.cs',
+    '**/*.java',
```

Comment:
Closing for now because I excluded them from this pr

Reply:
The Java API has a code formatter called spotless that runs as part of the test suite, if you apply this one it'll get very confused and fail all the CI pipelines with Java in them.

Reply:
Thanks for the info! I excluded java. Does the current change look good to you?

Reply:
Yes, though the Java binding includes some C code with the extension `.c`. You could apply the formatter to that, we don't currently do any formatting on the C code for the Java binding.

Reply:
> I think there may be some existing conventions for Java and C#, and also for WinML code. @Craigacp @yuslepukhin @fdwr Are you all fine with these formatting changes?

@edgchen1: Note I can't actually review this anyway because my browser keeps freezing due to the huge size of the change 😅.

My primary goal for any code formatting effort is *readability* 🧐 above blanket homogeneity, and automated code formatters are historically bulldozers that *sacrifice* readability with a simplistic set of rules 🐂🚧. If we are applying formatting more rigidly, I would highly value formatting rules that favor easy visual scannability of foreign logic and favor nicer diffing, neither of the which the current clang format ameliorates:

- it's much harder to diff the addition/removal/reordering/renaming of parameters in function definitions and function calls with ragged parameter wrap. Either putting all params on the same line when they fit, or breaking one param per line when they don't is linearly scannable and trivially diffable.
- the indentation logic is mystifying bouncing between 2 and 4 spaces in a bunch of cases. Discerning where the function parameters begin vs which line is the first executable statement is slower.
- following nested statements and logic is more challenging with tiny indents like 2.
- note I *do* value that ORT blesses up to 120 columns, because working some in TF and Chrome, the 80-column limit causes a lot of premature line wrapping which fragments lines and makes following the flow more difficult (especially in the year 2023 on my 4k monitor and multimon, where we're not limited to 640x480 from the 90's :b).

Reply:
The coding style of onnxruntime is stated in https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md . If some part of ONNX Runtime wants to use a different style, welcome to update the file and bring it into discussion. I think the most important thing is all of us should be on the same page.

Reply:
@fdwr Thanks for sharing your thoughts!

- I am very much in agreement with the first point. I am going to create an updated rule for discussion. This pr like Changming pointed out simply enforced the clang-format rules previously defined. I will play with the `AlignAfterOpenBracket` option for this. Please LMK if you have further suggestions.
- Regarding line length, I would encourage us to look at it from a perspective of inclusion. I have read from black's (the python formatter we use) documentation (https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length) that 90-ish can be an appropriate choice, because longer lines tend to be harder for visually impaired readers. Personally I work on a laptop monitor because larger screens hurt my eyes. Shorter lines make diff'ing easier when it's displayed side by side. Just my personal experience so cannot speak for everyone :)

Reply:
> Rules should be clear

@snnn: Yep, totally agreed that coding conventions should be documented, and that having generally consistent styles reduces cognitive burden - the maxim that when in the Rome part of code, mimic the Rome part of code. I wasn't implying we shouldn't have rules, but moreso I'm noting that sometimes we elevate one means of reducing cognitive burden (e.g. employing rigid rules or "make it all look the same" mantra or "make it look like what we inherited before from this other project") while forgetting the primary goal in any formatting recommendations is really to enable the reader to quickly and reliably ascertain what the new code is doing. Sometimes an *exception* to the rule actually makes the code more readable than the default rule. e.g. Sometimes you have a wider table full of values where keeping the values in each column aligned aids readability, but the formatting bulldozer comes along and resequences them all, introducing staggered columns. So, thankfully there is the `// clang-format off` option for such cases. 😁

> Please LMK if you have further suggestions.

@justinchuby: That's all I have. Thanks for listening.

Reply:
Created https://github.com/microsoft/onnxruntime/pull/15598/ for discussion

An example should reflect @fdwr 's first point:

![image](https://user-images.githubusercontent.com/11205048/233260977-cf940cfd-76a0-4fb1-9e27-5fc920b859e1.png)

Reply:
@justinchuby Did you intend to also rewrap enums?

![image](https://user-images.githubusercontent.com/1809166/233263984-dbcc8146-6721-453a-9ab8-6b75bbb95d46.png)

Back when I was playing around with a half dozen formatters, I found this online clang formatter very quick and easy to try out different settings: https://zed0.co.uk/clang-format-configurator/

Reply:
I think the rewrap was a side effect of clangformat seeing it has enough horizontal space to fit everything. I am going to consult with chatgpt

Reply:
Couldn't figure it out with the configurator or with chatgpt. Do you have suggestions?

Reply:
So we need a trailing comma.
![image](https://user-images.githubusercontent.com/11205048/233268596-7940f9d7-7e0a-4783-af76-a55e8f80512b.png)

Reply:
chatgpt helpfully created a fix script: https://github.com/microsoft/onnxruntime/pull/15606/files